File: Python/0100-same-tree.py
# time complexity: O(n)
# space complexity: O(n)
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        if p == None and q == None:
            return True
        if p == None or q == None:
            return False
        if p.val != q.val:
            return False
        return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)


p = TreeNode(1)
p.left = TreeNode(2)
p.right = TreeNode(2)

q = TreeNode(1)
q.left = TreeNode(2)
q.right = TreeNode(2)

print(Solution().isSameTree(p, q))

File: JavaScript/0100-same-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {boolean}
 */

var isSameTree = function (p, q) {
	if (p === null && q === null)
		return true;
	if (p === null || q === null)
		return false;
	if (p.val !== q.val)
		return false;
	return isSameTree(p.left, q.left) && isSameTree(p.right, q.right);
};

// Example usage:
var p = new TreeNode(1);
p.left = new TreeNode(2);
p.right = new TreeNode(2);

var q = new TreeNode(1);
q.left = new TreeNode(2);
q.right = new TreeNode(2);

console.log(isSameTree(p, q));
